File: 프로그래머스/2/62048. 멀쩡한 사각형/멀쩡한 사각형.js
function solution(w, h) {
    function getGCD(a, b) {
      if (b === 0) return a;
      return getGCD(b, a % b);
    }
    
    // 패턴의 개수
    const gcd = getGCD(w, h)
    const x = w / gcd
    const y = h / gcd
    const patternCount = w / x
    console.log(patternCount) // 4
    
    // 패턴이 차지하는 사각형 개수
    let unusable
    if (x === y) unusable = x
    else unusable = x + y - 1
    
    return w * h - patternCount * unusable
}
// 패턴의 종료지점
// 2, 3
// 4, 6
// 6, 9
// 8, 12 (+ 2, + 3씩 증가)
// 규칙 : W와 H의 최대공약수 4로 W와 H를 각각 나누면 몫이 각각 2, 3임

// 패턴이 차지하는 사각형 개수
// 패턴이 1*1 일 때 - 1개
// 패턴이 1*2 일 때 - 2개
// 패턴이 2*2 일 때 - 2개
// 패턴이 2*3 일 때 - 4개
// 패턴이 3*4 일 때 - 6개
// 규칙 : 정사각형일 때는 x개
//       직사각형일 때는 x + y - 1개

File: 프로그래머스/2/62048. 멀쩡한 사각형/멀쩡한 사각형.py
def solution(w,h):
    gcd = get_gcd(w, h)
    
    # 패턴의 가로 세로 길이
    pattern_w = w // gcd
    pattern_h = h // gcd
    pattern_count = w / pattern_w
    
    # 전체 사각형 개수 - 패턴이 차지하는 사각형 개수
    # 패턴이 차지하는 사각형 개수 규칙: 정사각형이면 x개, 직사각형이면 x + y - 1개
    unusable = 0
    if pattern_w == pattern_h:
        unusable = pattern_w
    else:
        unusable = pattern_w + pattern_h - 1
    
    return w * h - unusable * pattern_count

# 최대 공약수
def get_gcd(a, b):
    if a < b:
        a, b = b, a
        
    if b == 0:
        return a

    r = a % b
    return get_gcd(b, r)